torchao/csrc/cpu/paged_attention_kernel.cpp

```diff
@@ -0,0 +1,535 @@
+  for (auto partition_id = 0; partition_id < max_num_partitions;
+       partition_id++) {
+    for (auto head_id = 0; head_id < num_heads; head_id++) {
+      for (auto seq_id = 0; seq_id < num_seqs; seq_id++) {
+        auto context_len = context_lens_ptr[seq_id];
+        auto partition_start = partition_id * PARTITION_SIZE;
+        if (partition_start >= context_len)
+          continue;
+        auto partition_end =
+            std::min(partition_start + PARTITION_SIZE, context_len);
+        auto token_num = partition_end - partition_start;
+        auto block_num = (token_num + block_size - 1) / block_size;
+        auto logical_block_start = partition_start / block_size;
+        auto logical_block_end = logical_block_start + block_num;
+        auto need_update = block_num > 1;
+        auto kv_head_id = head_id / kv_head_group_size;
+        auto q_ptr_start = query_ptr + seq_id * q_strideN + head_id * q_strideH;
+        auto max_logits_offset = seq_id * max_logits_strideN +
+                                 head_id * max_logits_strideH + partition_id;
+        auto exp_sum_offset =
+            seq_id * exp_sum_strideN + head_id * exp_sum_strideH + partition_id;
+        auto tmp_out_start = tmp_out_ptr + seq_id * tmp_out_strideN +
+                             head_id * tmp_out_strideH +
+                             partition_id * tmp_out_strideS;
+        accum_t logits[PARTITION_SIZE] __attribute__((aligned(64))) = {0};
+        auto logits_position = 0;
+        // 1)calculate the matmul(query, key) for this partition
+        for (auto logical_block_id = logical_block_start;
+             logical_block_id < logical_block_end; logical_block_id++) {
+          auto physical_block_id =
+              block_tables_ptr[seq_id * max_num_blocks_per_seq +
+                               logical_block_id];
+          auto tokens_in_block =
+              std::min(block_size, context_len - logical_block_id * block_size);
+          auto token_start = logical_block_id * block_size;
+          auto token_end = token_start + tokens_in_block;
+          for (auto token_id = token_start; token_id < token_end; token_id++) {
+            auto block_offset = token_id - token_start;
+            auto k_cache_start =
+                key_cache_ptr + physical_block_id * kv_block_strideN +
+                block_offset * kv_block_strideP + kv_head_id * kv_block_strideH;
+            reduce_head<scalar_t, accum_t>(q_ptr_start, k_cache_start,
+                                           &(logits[logits_position]),
+                                           head_size);
+            logits_position++;
+          }
+        }
+        // 2) calculate the max and exp_sum for this partition
+        auto partition_max = -std::numeric_limits<accum_t>::infinity();
+        if (has_attn_mask) {
+          _scale_attn_mask_fusion_kernel<accum_t, accum_t>(
+              logits,
+              attn_mask_ptr + seq_id * mStrideB + head_id * mStrideH +
```

Comment:
I guess we need to carefully handle the case where the size is 1 in some dim of the mask here.